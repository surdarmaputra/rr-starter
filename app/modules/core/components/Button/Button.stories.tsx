import type { Meta, StoryObj } from '@storybook/react';

import { Button, type ButtonProps } from './Button';

type Story = StoryObj<typeof Button>;

const meta: Meta = {
  component: Button,
  title: 'Core/Components/Button',
};

export default meta;

export const Default: Story = {
  args: {
    children: 'Button',
  },
  render: (args: ButtonProps) => <Button {...args} />,
};
