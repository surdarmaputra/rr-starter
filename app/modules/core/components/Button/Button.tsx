import { Slot } from '@radix-ui/react-slot';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

import type { TestableComponentProps } from '../../types';

const buttonVariants = cva(
  `
    inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold
    ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2
    focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none
    disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0
    dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300
  `,
  {
    variants: {
      variant: {
        primary: 'bg-primary text-white hover:bg-primary-hover',
        'primary-outlined':
          'border border-primary text-primary hover:bg-primary-hover hover:text-white',
        'primary-ghost': 'text-primary hover:bg-primary-hover hover:text-white',
        secondary: 'bg-secondary text-white hover:bg-secondary-hover',
        'secondary-outlined':
          'border border-secondary text-secondary hover:bg-secondary-hover hover:text-white',
        'secondary-ghost':
          'text-secondary hover:bg-secondary-hover hover:text-white',
        success: 'bg-success text-white hover:bg-success-hover',
        'success-outlined':
          'border border-success text-success hover:bg-success-hover hover:text-white',
        'success-ghost': 'text-success hover:bg-success-hover hover:text-white',
        warning: 'bg-warning text-white hover:bg-warning-hover',
        'warning-outlined':
          'border border-warning text-warning hover:bg-warning-hover hover:text-white',
        'warning-ghost': 'text-warning hover:bg-warning-hover hover:text-white',
        danger: 'bg-danger text-white hover:bg-danger-hover',
        'danger-outlined':
          'border border-danger text-danger hover:bg-danger-hover hover:text-white',
        'danger-ghost': 'text-danger hover:bg-danger-hover hover:text-white',
        link: 'text-textTitle underline-offset-4 hover:underline',
      },
      size: {
        default: 'h-10 px-4 py-2',
        sm: 'h-9 px-3',
        lg: 'h-11 px-8',
        icon: 'h-10 w-10',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'default',
    },
  },
);

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants>,
    TestableComponentProps {
  asChild?: boolean;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : 'button';
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      />
    );
  },
);

Button.displayName = 'Button';

export { Button, buttonVariants };
