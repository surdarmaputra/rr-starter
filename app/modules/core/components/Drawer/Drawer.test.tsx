import { render, screen, within } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { Button } from '../Button/Button';
import {
  Drawer,
  DrawerClose,
  DrawerContent,
  DrawerDescription,
  DrawerFooter,
  DrawerHeader,
  type DrawerProps,
  DrawerTitle,
} from './Drawer';

describe('Drawer', () => {
  function renderComponent(props?: DrawerProps) {
    return render(
      <Drawer {...props}>
        <DrawerContent data-testid="DrawerContent-testid">
          <DrawerHeader data-testid="DrawerHeader-testid">
            <DrawerTitle>Are you absolutely sure?</DrawerTitle>
            <DrawerDescription>Header description</DrawerDescription>
          </DrawerHeader>
          <div data-testid="DrawerContentBody-testid">Content body</div>
          <DrawerFooter data-testid="DrawerFooter-testid">
            <Button>Save changes</Button>
            <DrawerClose asChild>
              <Button variant="primary-outlined">Cancel</Button>
            </DrawerClose>
          </DrawerFooter>
        </DrawerContent>
      </Drawer>,
    );
  }

  test('hides content when not opened', async () => {
    renderComponent();
    expect(
      screen.queryByTestId('DrawerContent-testid'),
    ).not.toBeInTheDocument();
  });

  test('displays content when opened', async () => {
    renderComponent({
      open: true,
    });
    const dialog = await screen.findByTestId('DrawerContent-testid');

    const header = within(dialog).getByTestId('DrawerHeader-testid');
    within(header).getByText(/Are you absolutely sure?/);
    within(header).getByText(/Header description/);

    const body = within(dialog).getByTestId('DrawerContentBody-testid');
    within(body).getByText(/Content body/);

    const footer = within(dialog).getByTestId('DrawerFooter-testid');
    within(footer).getByRole('button', { name: /Save changes/ });
  });

  test('renders snapshot', async () => {
    renderComponent({
      open: true,
    });
    const dialog = await screen.findByTestId('DrawerContent-testid');
    expect(dialog).toMatchSnapshot();
  });
});
