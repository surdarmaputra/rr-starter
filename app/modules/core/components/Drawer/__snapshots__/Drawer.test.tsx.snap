// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Drawer > renders snapshot 1`] = `
<div
  aria-describedby="radix-:r8:"
  aria-labelledby="radix-:r7:"
  class="fixed inset-x-0 bottom-0 z-50 mt-24 flex h-auto flex-col rounded-t-[10px] border border-slate-200 bg-white dark:border-slate-800 dark:bg-slate-950"
  data-state="open"
  data-testid="DrawerContent-testid"
  data-vaul-animate="true"
  data-vaul-custom-container="false"
  data-vaul-delayed-snap-points="false"
  data-vaul-drawer=""
  data-vaul-drawer-direction="bottom"
  data-vaul-snap-points="false"
  id="radix-:r6:"
  role="dialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <div
    class="mx-auto mt-4 h-2 w-[100px] rounded-full bg-slate-100 dark:bg-slate-800"
  />
  <div
    class="grid gap-1.5 p-4 text-center sm:text-left"
    data-testid="DrawerHeader-testid"
  >
    <h2
      class="text-lg font-semibold leading-none tracking-tight"
      id="radix-:r7:"
    >
      Are you absolutely sure?
    </h2>
    <p
      class="text-sm text-slate-500 dark:text-slate-400"
      id="radix-:r8:"
    >
      Header description
    </p>
  </div>
  <div
    data-testid="DrawerContentBody-testid"
  >
    Content body
  </div>
  <div
    class="mt-auto flex flex-col gap-2 p-4"
    data-testid="DrawerFooter-testid"
  >
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-primary text-white hover:bg-primary-hover h-10 px-4 py-2"
    >
      Save changes
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 border border-primary text-primary hover:bg-primary-hover hover:text-white h-10 px-4 py-2"
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;
